### DISCUSSION TITLE
Handling optional attributes in SQLAlchemy Declarative Mapping abstract classes with dataclass

### DISCUSSION THREAD

**Discussion by oliviernguyenquoc** [sqlalchemy/sqlalchemy#10027] [Usage Questions] [answered]
## Description
I'm currently trying to implement SQLAlchemy's [Declarative Mapping with Pydantic's dataclass](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#integrating-with-alternate-dataclass-providers-such-as-pydantic) with SQLAlchemy 2.0 in an Async context. However, I encountered a use case that I'm unable to solve with my existing `Base` class. 

## Objective
The workflow that I'm trying to achieve is as follows:
1. A POST request is made to my API.
2. The input format is verified.
3. An output message with additional attribute is sent.

The challenge lies in handling the optional attribute `attribute1` in my input message, which defaults to `False` when not provided.

## Problem
The current implementation does **not handle `attribute1` as expected**. When `asdict` is called on `b_input` in `b_output = BOutput(**asdict(b_input))`, `attribute1` seems to be misinterpreted. As a result, the following output is returned:

`print(b_input)` results in `BInput(name='Test B 1', attribute1=<sqlalchemy.orm.properties.MappedColumn object at 0x7f3a86f1e8c0>)`

However, what I expect to see is:
`BInput(name='Test B 1', attribute1=False)`

## Question
Is there an alternative approach to `attribute1: Mapped[bool] = mapped_column(Boolean(), default=False)`?

## Environment and Setup
The issue can be reproduced in the following environment:
- SQLAlchemy 2.0.16
- Fastapi  0.100.0-beta1
- Pydantic 2.0b3

The relevant model and API are as follows:

My model:
```python
class Base(
    MappedAsDataclass,
    DeclarativeBase,
    dataclass_callable=pydantic.dataclasses.dataclass,
):
    class Config:
        orm_mode = True


class BInput(Base):
    __abstract__ = True
    name: Mapped[str] = mapped_column(String(50))
    attribute1: Mapped[bool] = mapped_column(Boolean(), default=False)


class BOutput(BInput):
    __tablename__ = "b"

    attribute2: Mapped[int] = mapped_column(Integer(), default=100)
    computed_info_1: int = 0
    computed_info_2: str = ""
```

My API:
```python
@router.post("", response_model=BOutput, status_code=201)
async def add(*, db: AsyncSession = Depends(get_db), b_input: BInput):
    b_output = BOutput(**asdict(b_input))
    b_output.computed_info_2 = "Some custom logic too add here"
    db.add(b_output)
    await db.commit()

    return b_output
```

## Test Case
I've also provided a test case to help reproduce the issue:

```python
@pytest.mark.asyncio
async def test_create_b(db_session: AsyncSession, client: AsyncClient):
    response = await client.post(
        "/v1/b",
        json={
            "name": "Test B 1",
        },
    )

    assert response.status_code == 201

    pack = response.json()
    print(pack)

    assert pack["name"] == "Test B 1"
    assert pack["attribute1"] == False
    assert pack["attribute2"] == 100
```

## Error
The error that occurs during execution is as follows:

<details>
  <summary>Python stack traceback</summary>
  
  ```
  /usr/local/lib/python3.11/site-packages/httpx/_client.py:1848: in post
    return await self.request(
/usr/local/lib/python3.11/site-packages/httpx/_client.py:1530: in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
/usr/local/lib/python3.11/site-packages/httpx/_client.py:1617: in send
    response = await self._send_handling_auth(
/usr/local/lib/python3.11/site-packages/httpx/_client.py:1645: in _send_handling_auth
    response = await self._send_handling_redirects(
/usr/local/lib/python3.11/site-packages/httpx/_client.py:1682: in _send_handling_redirects
    response = await self._send_single_request(request)
/usr/local/lib/python3.11/site-packages/httpx/_client.py:1719: in _send_single_request
    response = await transport.handle_async_request(request)
/usr/local/lib/python3.11/site-packages/httpx/_transports/asgi.py:162: in handle_async_request
    await self.app(scope, receive, send)
/usr/local/lib/python3.11/site-packages/fastapi/applications.py:281: in __call__
    await super().__call__(scope, receive, send)
/usr/local/lib/python3.11/site-packages/starlette/applications.py:122: in __call__
    await self.middleware_stack(scope, receive, send)
/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py:184: in __call__
    raise exc
/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py:162: in __call__
    await self.app(scope, receive, _send)
/usr/local/lib/python3.11/site-packages/starlette/middleware/exceptions.py:79: in __call__
    raise exc
/usr/local/lib/python3.11/site-packages/starlette/middleware/exceptions.py:68: in __call__
    await self.app(scope, receive, sender)
/usr/local/lib/python3.11/site-packages/fastapi/middleware/asyncexitstack.py:20: in __call__
    raise e
/usr/local/lib/python3.11/site-packages/fastapi/middleware/asyncexitstack.py:17: in __call__
    await self.app(scope, receive, send)
/usr/local/lib/python3.11/site-packages/starlette/routing.py:718: in __call__
    await route.handle(scope, receive, send)
/usr/local/lib/python3.11/site-packages/starlette/routing.py:276: in handle
    await self.app(scope, receive, send)
/usr/local/lib/python3.11/site-packages/starlette/routing.py:66: in app
    response = await func(request)
/usr/local/lib/python3.11/site-packages/fastapi/routing.py:273: in app
    raw_response = await run_endpoint_function(
/usr/local/lib/python3.11/site-packages/fastapi/routing.py:190: in run_endpoint_function
    return await dependant.call(**values)
app/router.py:24: in add
    b_output = BOutput(**asdict(b_input))
/usr/local/lib/python3.11/dataclasses.py:1277: in asdict
    return _asdict_inner(obj, dict_factory)
/usr/local/lib/python3.11/dataclasses.py:1284: in _asdict_inner
    value = _asdict_inner(getattr(obj, f.name), dict_factory)
/usr/local/lib/python3.11/dataclasses.py:1318: in _asdict_inner
    return copy.deepcopy(obj)
/usr/local/lib/python3.11/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
/usr/local/lib/python3.11/copy.py:271: in _reconstruct
    state = deepcopy(state, memo)
/usr/local/lib/python3.11/copy.py:146: in deepcopy
    y = copier(x, memo)
/usr/local/lib/python3.11/copy.py:211: in _deepcopy_tuple
    y = [deepcopy(a, memo) for a in x]
/usr/local/lib/python3.11/copy.py:211: in <listcomp>
    y = [deepcopy(a, memo) for a in x]
/usr/local/lib/python3.11/copy.py:146: in deepcopy
    y = copier(x, memo)
/usr/local/lib/python3.11/copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
/usr/local/lib/python3.11/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
/usr/local/lib/python3.11/copy.py:271: in _reconstruct
    state = deepcopy(state, memo)
/usr/local/lib/python3.11/copy.py:146: in deepcopy
    y = copier(x, memo)
/usr/local/lib/python3.11/copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
/usr/local/lib/python3.11/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
/usr/local/lib/python3.11/copy.py:271: in _reconstruct
    state = deepcopy(state, memo)
/usr/local/lib/python3.11/copy.py:146: in deepcopy
    y = copier(x, memo)
/usr/local/lib/python3.11/copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
/usr/local/lib/python3.11/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
/usr/local/lib/python3.11/copy.py:271: in _reconstruct
    state = deepcopy(state, memo)
/usr/local/lib/python3.11/copy.py:146: in deepcopy
    y = copier(x, memo)
/usr/local/lib/python3.11/copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
/usr/local/lib/python3.11/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
/usr/local/lib/python3.11/copy.py:271: in _reconstruct
    state = deepcopy(state, memo)
/usr/local/lib/python3.11/copy.py:146: in deepcopy
    y = copier(x, memo)
/usr/local/lib/python3.11/copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
/usr/local/lib/python3.11/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
/usr/local/lib/python3.11/copy.py:271: in _reconstruct
    state = deepcopy(state, memo)
/usr/local/lib/python3.11/copy.py:146: in deepcopy
    y = copier(x, memo)
/usr/local/lib/python3.11/copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
/usr/local/lib/python3.11/copy.py:146: in deepcopy
    y = copier(x, memo)
/usr/local/lib/python3.11/copy.py:206: in _deepcopy_list
    append(deepcopy(a, memo))
/usr/local/lib/python3.11/copy.py:146: in deepcopy
    y = copier(x, memo)
/usr/local/lib/python3.11/copy.py:211: in _deepcopy_tuple
    y = [deepcopy(a, memo) for a in x]
/usr/local/lib/python3.11/copy.py:211: in <listcomp>
    y = [deepcopy(a, memo) for a in x]
/usr/local/lib/python3.11/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
/usr/local/lib/python3.11/copy.py:271: in _reconstruct
    state = deepcopy(state, memo)
/usr/local/lib/python3.11/copy.py:146: in deepcopy
    y = copier(x, memo)
/usr/local/lib/python3.11/copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
/usr/local/lib/python3.11/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
/usr/local/lib/python3.11/copy.py:271: in _reconstruct
    state = deepcopy(state, memo)
/usr/local/lib/python3.11/copy.py:146: in deepcopy
    y = copier(x, memo)
/usr/local/lib/python3.11/copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
/usr/local/lib/python3.11/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
/usr/local/lib/python3.11/copy.py:271: in _reconstruct
    state = deepcopy(state, memo)
/usr/local/lib/python3.11/copy.py:146: in deepcopy
    y = copier(x, memo)
/usr/local/lib/python3.11/copy.py:211: in _deepcopy_tuple
    y = [deepcopy(a, memo) for a in x]
/usr/local/lib/python3.11/copy.py:211: in <listcomp>
    y = [deepcopy(a, memo) for a in x]
/usr/local/lib/python3.11/copy.py:146: in deepcopy
    y = copier(x, memo)
/usr/local/lib/python3.11/copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
/usr/local/lib/python3.11/copy.py:161: in deepcopy
    rv = reductor(4)
/usr/local/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:1311: in __getattr__
    value = getattr(self, f"_memoized_attr_{key}")()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sqlalchemy.event.attr._JoinedListener object at 0x7f30a7847760>

    def _memoized_attr_ref(self) -> weakref.ref[RefCollection[_ET]]:
>       return weakref.ref(self, registry._collection_gced)
E       TypeError: cannot create weak reference to '_JoinedListener' object

/usr/local/lib/python3.11/site-packages/sqlalchemy/event/attr.py:81: TypeError
  ```
</details>

If required, I can provide a minimal reproducible example which will be quite similar to the one I've provided here. 

Looking forward to any insights and guidance that might help me resolve this issue.
Tell me if I need to make an issue instead

**Update:**

I've clarified and updated the original issue to ease troubleshooting

**Comment by CaselIT** [comment]
Hi,

I'm not sure what's at fault. Trying to create a runnable example from your snipped does not seem to result in any error. Try to modify it to make it return an error

EDIT: updated script to reproduce. This happens also with plain dataclasses

```py
from sqlalchemy.orm import (
    DeclarativeBase,
    MappedAsDataclass,
    Mapped,
    mapped_column,
)
from sqlalchemy import Boolean, Integer, String
import dataclasses as DC
import pydantic

class Base(
    MappedAsDataclass,
    DeclarativeBase,
#    dataclass_callable=pydantic.dataclasses.dataclass,
):
    class Config:
        orm_mode = True

class BInput(Base):
    __abstract__ = True
    id: Mapped[int] = mapped_column(Integer, primary_key=True, init=False)
    name: Mapped[str] = mapped_column(String(50))
    attribute1: Mapped[bool] = mapped_column(Boolean(), default=False)

class BOutput(BInput):
    __tablename__ = "b"

    attribute2: Mapped[int] = mapped_column(Integer(), default=100)
    computed_info_1: int = 0
    computed_info_2: str = ""

b_input = BInput(name="foo")
print(DC.asdict(b_input))
b_output = BOutput(**DC.asdict(b_input))
b_output.attribute2 = 42
print(DC.asdict(b_output))
```

**Reply by oliviernguyenquoc** [reply]
You specified `b_input = BInput(name="foo", attribute1=True)`
I need that code to work with either:
1. `b_input = BInput(name="foo")`with attribute1 to False by default (and that is not working on my side) 
2. `b_input = BInput(name="foo", attribute1=True)`

**Reply by CaselIT** [reply]
oh this may be caused by how dataclasses sets default values here, since they are placed on the class dict.

But when a class is mapped by sqlalchemy those defaults are removed entirely, and you have a descriptor there.

I'll try using `b_input = BInput(name="foo")`

**Reply by CaselIT** [reply]
This seems indeed a bug.
Printing the following also returns unexpected values
```py
b_input = BInput(name="foo")
print(b_input.id, b_input.attribute1)
print(b_input.__dict__)
```
```
<sqlalchemy.orm.properties.MappedColumn object at 0x00000154161CF0A0> <sqlalchemy.orm.properties.MappedColumn object at 0x00000154161CF2E0>
{'id': <sqlalchemy.orm.properties.MappedColumn object at 0x00000154161CF0A0>, 'name': 'foo', 'attribute1': <sqlalchemy.orm.properties.MappedColumn object at 0x00000154161CF2E0>}
```

The issue here seems duet to `__abstract__ = True`.
cc @zzzeek

**Reply by zzzeek** [reply]
it's not valid to instantiate BInput, so that would not be a "bug".   

I get a stack trace here for some reason also 

```py
$ python test3.py 
Traceback (most recent call last):
  File "/mnt/photon_home/classic/dev/sqlalchemy/test3.py", line 35, in <module>
    print(DC.asdict(b_input))
          ^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/dataclasses.py", line 1277, in asdict
    return _asdict_inner(obj, dict_factory)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/dataclasses.py", line 1284, in _asdict_inner
    value = _asdict_inner(getattr(obj, f.name), dict_factory)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/dataclasses.py", line 1318, in _asdict_inner
    return copy.deepcopy(obj)
           ^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 271, in _reconstruct
    state = deepcopy(state, memo)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 146, in deepcopy
    y = copier(x, memo)
        ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 211, in _deepcopy_tuple
    y = [deepcopy(a, memo) for a in x]
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 211, in <listcomp>
    y = [deepcopy(a, memo) for a in x]
         ^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 146, in deepcopy
    y = copier(x, memo)
        ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
                             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 271, in _reconstruct
    state = deepcopy(state, memo)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 146, in deepcopy
    y = copier(x, memo)
        ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
                             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 271, in _reconstruct
    state = deepcopy(state, memo)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 146, in deepcopy
    y = copier(x, memo)
        ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
                             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 271, in _reconstruct
    state = deepcopy(state, memo)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 146, in deepcopy
    y = copier(x, memo)
        ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
                             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 271, in _reconstruct
    state = deepcopy(state, memo)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 146, in deepcopy
    y = copier(x, memo)
        ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
                             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 271, in _reconstruct
    state = deepcopy(state, memo)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 146, in deepcopy
    y = copier(x, memo)
        ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 211, in _deepcopy_tuple
    y = [deepcopy(a, memo) for a in x]
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 211, in <listcomp>
    y = [deepcopy(a, memo) for a in x]
         ^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 146, in deepcopy
    y = copier(x, memo)
        ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
                             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/copy.py", line 161, in deepcopy
    rv = reductor(4)
         ^^^^^^^^^^^
  File "/mnt/photon_home/classic/dev/sqlalchemy/lib/sqlalchemy/util/langhelpers.py", line 1311, in __getattr__
    value = getattr(self, f"_memoized_attr_{key}")()
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/photon_home/classic/dev/sqlalchemy/lib/sqlalchemy/event/attr.py", line 81, in _memoized_attr_ref
    return weakref.ref(self, registry._collection_gced)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: cannot create weak reference to '_JoinedListener' object

```

this is weird?

**Reply by CaselIT** [reply]
asdict is trying to copy a mapped columns and failing

**Reply by oliviernguyenquoc** [reply]
Got the same error

**Comment by zzzeek** [comment]
> The current implementation does not handle attribute1 as expected. When asdict is called on b_input in b_output = BOutput(**asdict(b_input)), attribute1 seems to be misinterpreted. As a result, the following output is returned:

> print(b_input) results in BInput(name='Test B 1', attribute1=<sqlalchemy.orm.properties.MappedColumn object at 0x7f3a86f1e8c0>)

> However, what I expect to see is:
> BInput(name='Test B 1', attribute1=False)


There's no ORM constructor or ORM-handling of these attributes on the abstract class because it's not valid to instantiate an abstract class:

https://en.wikipedia.org/wiki/Abstract_type

> In [programming languages](https://en.wikipedia.org/wiki/Programming_languages), an abstract type is a type in a [nominative type system](https://en.wikipedia.org/wiki/Nominative_type_system) that cannot be [instantiated](https://en.wikipedia.org/wiki/Instance_(computer_science)) directly; a type that is not abstract – which can be instantiated – is called a concrete type.

SQLAlchemy isn't adding special steps here to prevent the instantiation but I guess we should.

**Reply by oliviernguyenquoc** [reply]
It's a valid point that I should not instantiate a abstract class. The thing is, I try to find a way **not** to persist this class in the database. It's the only way I find to do it (and in the same time be able to have this optional attribute1 mecanism)

**Reply by CaselIT** [reply]
Why aren't you using the output class directly?
What's the use case for having an input class that then gets converted to an output one?

**Reply by oliviernguyenquoc** [reply]
By using [FastAPI](https://fastapi.tiangolo.com/), we are able to check the input and the ouput with Pydantic, by design, by this API framework.
So, the input is automatically parsed as a `BInput`by the framework with pydantic, then we add some logic and return a different schema as an output.

**Reply by oliviernguyenquoc** [reply]
@CaselIT Any idea ? Should I create an issue instead ?

**Reply by zzzeek** [reply]
we would make "abstract" classes raise an error here when instantiated, the dataclass + mapped + `__abstract__` combination here will never have a constructor that acts like that of a mapped class because the class is not mapped or instrumented.

**Reply by CaselIT** [reply]
> @CaselIT Any idea ? Should I create an issue instead ?

best option is probably to use a different class, that's not related to the one used by sqlalchemy if you want two different classes, one for the input and one that's mapped

**Reply by oliviernguyenquoc** [reply]
So, if I understand you well, it's not possible to have a unique class to define my types (pydantic) and define my table definition (sqlalchemy).

I though one of the main goal with Declarative Mapping was to be able to do such thing and avoid usage of package like [sqlmodel](https://www.sqlalchemy.org/blog/2023/01/26/sqlalchemy-2.0.0-released/).

From [SQLAlchemy 2.0.0 Released](https://www.sqlalchemy.org/blog/2023/01/26/sqlalchemy-2.0.0-released/) note:
"All new plugin-free pep-484 compatible ORM syntaxes - The ORM Declarative style of mapper configuration now has an all new look, borrowing heavily from systems such as Python dataclasses and SQLModel to provide for [Annotation-driven ORM declarations](https://docs.sqlalchemy.org/en/latest/changelog/whatsnew_20.html#orm-declarative-models), using runtime interpretation of pep-484 annotations to produce mapped classes that are fully typed and compatible with any type checker or IDE."

Maybe I missunderstood the intention of it. Can you confirm ?

**Reply by zzzeek** [reply]
The intention of ORM Annotated Declarative is so that you can make SQLAlchemy models that will pass fully with type checkers such as mypy and pyright, without the use of plugins.    It borrows from styles that can be seen in SQLModel and Pydantic to achieve this.

>  it's not possible to have a unique class to define my types (pydantic) and define my table definition (sqlalchemy).

i dont think I'm understanding what you mean.  SQLAlchemy's Declarative feature allows one to define a class and its associated table definition at the same time.    As far as if you are asking if we integrate with pydantic, we allow a pydantic's dataclasses function to be integrated with SQLAlchemy's "dataclasses" integration as documented here: https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#integrating-with-alternate-dataclass-providers-such-as-pydantic .  This is known to work best with Pydantic 2.0 however we don't offer explicit support for pydantic.

overall I dont see how this has anything to do with being able to instantiate abstract classes.

**Reply by oliviernguyenquoc** [reply]
Sorry if that was not clear enough.

The issue is that I can't write a SQLAlchemy class without specifying __tablename__. So the only (bad) solution is to specify an abstract class, which is not sutaible either.

What I wanted to do is to have `BInput` to be my pydantic type for input **and**  a class with fields, that need to be heritated to define my table "b" table.

What would have been so cool is to be able to write Declerative Mapping style class without specifying __tablename__ (so with no table generation).

Maybe I'm over-engineering too much for nothing and need to work to improve [SQLModel](https://github.com/tiangolo/sqlmodel) compability to do such a thing.

**Reply by CaselIT** [reply]
> So, if I understand you well, it's not possible to have a unique class to define my types (pydantic) and define my table definition (sqlalchemy).

That's possible. The issue here is that you are trying to use an abstract class from sqlalchemy point of view as your types. If you were to use `BOutput` everything would be fine.

> What I wanted to do is to have BInput to be my pydantic type for input and a class with fields, that need to be heritated to define my table "b" table.

You may be able to fiddle around with the dataclass constructor function, by wrapping the pydantic one in your own to present a different class state to the real dataclass function.

Using default_factory instead of default may also improve things

**Reply by oliviernguyenquoc** [reply]
On second thoughts, abstract class is a false debate.

The issue is still the same if I write this code:

```python

from sqlalchemy.orm import Mapped, MappedAsDataclass, mapped_column

class Base(
    MappedAsDataclass,
    DeclarativeBase,
#    dataclass_callable=pydantic.dataclasses.dataclass,
):
    class Config:
        orm_mode = True


class BInput(MappedAsDataclass):
    id: Mapped[int] = mapped_column(Integer, primary_key=True, init=False)
    name: Mapped[str] = mapped_column(String(50))
    attribute1: Mapped[bool] = mapped_column(Boolean(), default=False)

class BOutput(Base, BInput):
    __tablename__ = "b"

    attribute2: Mapped[int] = mapped_column(Integer(), default=100)
    computed_info_1: int = 0
    computed_info_2: str = ""

b_input = BInput(name="foo", attribute1=False)
print(DC.asdict(b_input))
b_output = BOutput(**DC.asdict(b_input))
b_output.attribute2 = 42
print(DC.asdict(b_output))

```

In this case, I have a bug on `DC.asdict(b_input)`:
`b_input` have this shape: `BInput(name="foo", attribute1=<sqlalchemy.orm.properties.MappedColumn object at 0x7f61cf4fa710>)` with the same weird error stack [you mentionned here previously](https://github.com/sqlalchemy/sqlalchemy/discussions/10027#discussioncomment-6318879).


IMO, at least there is something to change in exception sent.


Anyway, thanks for your support. Really appreciated for your patience and SQLAlchemy is still a gem in the python echosystem ;)

PS @CaselIT: Still trying to apply your advices but seems taht`default_factory`is not helping that much

**Reply by zzzeek** [reply]
sure, that's a mixin, you can "instantiate" that but SQLAlchemy does not map it, so the mapped_column() attributes are not removed and there is no solution to that problem here other than setting up init=False on the class.

@CaselIT didn't we have some issue where we were going to force init=False for some subset of dataclasses ?

**Reply by CaselIT** [reply]
> @CaselIT didn't we have some issue where we were going to force init=False for some subset of dataclasses ?

I remember that point, but maybe it was this issue here. IIRC the idea was that if a class is abstract we were to wrap init to warn on instantiation and/or force init=False.

> PS @CaselIT: Still trying to apply your advices but seems tahtdefault_factoryis not helping that much

```py
class Id(MappedAsDataclass):
    id: Mapped[int] = mapped_column(Integer, primary_key=True, init=False)

class BInput(Base):
    __abstract__ = True
    name: Mapped[str] = mapped_column(String(50))
    attribute1: Mapped[bool] = mapped_column(Boolean(), default_factory=lambda: False)

class BOutput(BInput, Id):
    __tablename__ = "b"

    attribute2: Mapped[int] = mapped_column(Integer(), default=100)
    computed_info_1: int = 0
    computed_info_2: str = ""
```
works in the example below. that said the fact that it's working is incidental, and probably not sorething sqlalchemy should commit to support

**Comment by zzzeek** [comment]
> Sorry if that was not clear enough.
> 
> The issue is that I can't write a SQLAlchemy class without specifying **tablename**. So the only (bad) solution is to specify an abstract class, which is not sutaible either.

A SQLAlchemy class is mapped to a database table, always, either directly or via inheritance.  that's their purpose.   if you want to instantiate classes that are not mapped to database tables, then you would not make that class a SQLAlchemy mapped class.   So make a separate class that does the business case you need and don't map it (that is, does not extend from the declarative base at all).     A mixin class satisfies this need as well.

> 
> What I wanted to do is to have `BInput` to be my pydantic type for input **and** a class with fields, that need to be heritated to define my table "b" table.

right but you are putting mapped_column() directives on that class, which doesnt make sense if you want to instantiate that class and have it not be associated with any database table.       pydantic has some kind of "two class approach" to this sort of thing, apparently.

> 
> What would have been so cool is to be able to write Declerative Mapping style class without specifying **tablename** (so with no table generation).

a mixin or abstract class can have this, but they are not mapped, so instantiation makes no sense.  instantiation means the class is managed from a persistence perspective and requires a Mapper.  The Mapper does not have a "exist, but we are not really mapped to any table" mode - that would be extremely complicated without there being a real use case.   the mapped_column() directives are not actual instance-level constructs, when the class is mapped, these get replaced with Python descriptors that do the real work for instances.


> 
> Maybe I'm over-engineering too much for nothing and need to work to improve [SQLModel](https://github.com/tiangolo/sqlmodel) compability to do such a thing.

what seems awkward here is that you are going for "classes that aren't mapped", which is a standard "high degree of separation of concerns" approach, that's OK, but then you want that same class to be used in a mapped hierarchy, which seems to defeat the whole purpose of separation of concerns.

you might want to look into imperative mappings, either declarative with imperative table, or full imperative, if you are going for "the datamodel classes and how they are persisted in a database should be separate concerns" model.

**Comment by zzzeek** [comment, marked as answer]
added #10064.  Instantiation of any MappedAsDataclass class that is not mapped is not supported by SQLAlchemy, and behavior is currently undefined.
